rustls/src/msgs/handshake.rs

```diff
@@ -1182,6 +1177,54 @@ impl HelloRetryRequest {
             _ => None,
         }
     }
+
+    fn payload_encode(&self, bytes: &mut Vec<u8>, purpose: Purpose) {
+        self.legacy_version.encode(bytes);
+        HELLO_RETRY_REQUEST_RANDOM.encode(bytes);
+        self.session_id.encode(bytes);
+        self.cipher_suite.encode(bytes);
+        Compression::Null.encode(bytes);
+
+        match purpose {
+            // Standard encoding encodes extensions as they appear.
+            Purpose::Standard => {
+                self.extensions.encode(bytes);
+            }
+            // For the purpose of ECH confirmation, the Encrypted Client Hello extension
+            // must have its payload replaced by 8 zero bytes.
+            //
+            // See draft-ietf-tls-esni-17 7.2.1:
+            // <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#name-sending-helloretryrequest-2>
+            // TODO(XXX): Replace with RFC reference once published.
+            Purpose::EchConfirmation => {
+                let extensions = LengthPrefixedBuffer::new(ListLength::U16, bytes);
+                for ext in &self.extensions {
+                    match ext.get_type() {
+                        ExtensionType::EncryptedClientHello => {
+                            HelloRetryExtension::EchHelloRetryRequest(vec![0u8; 8])
+                                .encode(extensions.buf);
+                        }
+                        _ => {
+                            ext.encode(extensions.buf);
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// The purpose a payload is being encoded for.
+///
+/// In some cases a handshake message may be encoded differently depending on the purpose
+/// the encoded message is being used for. For example, a [ServerHelloPayload] may be encoded
+/// with the last 8 bytes of the random zeroed out when being encoded for ECH confirmation.
+pub(crate) enum Purpose {
```

Comment:
> Naming nit: not sure I love the Purpose name here... HandshakeMode/HandshakeKind/Context? 

I'm not deeply attached to it :-) `Context` seems reasonable to me. I'm less keen on the "Handshake" variants because we use both types of encoding in the context of a handshake where ECH has been offered and I think this is more of a per-message idea.

>  Does the standard use a word for distinguishing between non-ECH and ECH handshakes?

I don't think so. E.g. in [the section](https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#backend-server) relevant for the server hello accept confirmation it describes "computing" the ServerHello. Ditto for the HRR version in 7.2.1 "The backend server begins by computing HelloRetryRequest as usual, except  ..."

Reply:
The downside of `Context` is that it is very generic and also used for example in the type passed to `Future::poll()`.

Reply:
> The downside of Context is that it is very generic

That's true. Maybe `Encoding::Standard` and `Encoding::EchConfirmation`?

Reply:
Yes, that seems better.

Reply:
Fixed.